src/main/kotlin/org/rust/lang/utils/RsDiagnostic.kt

```diff
@@ -864,6 +866,23 @@ sealed class RsDiagnostic(
         }
     }
 
+    class ImplForNonAdtError(
+        element: PsiElement,
+        private val impl: RsImplItem,
+        private val type: Ty
+    ) : RsDiagnostic(element) {
+        override fun prepare() = PreparedAnnotation(
+            ERROR,
+            E0118,
+            "Can impl only `struct`s, `enum`s and `union`s (and primitives in `no_core` crates)",
```

Comment:
I don't think it's reasonable to mention primitives here, it isn't useful for must users. Mention trait objects looks much more reasonable